### PR TITLE
Support for HTTP/2 (server-side)

### LINKED ISSUE #1953: Bump the main branch to 3.0
**Is your feature request related to a problem?**
We're working on few changes in cross cluster replication for 3.0. For us to be able to use CCR on 3.0, we need to have security plugin also bumped up to 3.0. 
I see that was already done as part of [this PR](https://github.com/opensearch-project/security/pull/1890) but was reverted [here](https://github.com/opensearch-project/security/pull/1901).  Not sure why it was reverted though.

**What solution would you like?**
Bump the main branch to 3.0 and add to OS build manifest.
cc: @DarshitChanpura @peternied @cliu123

### PR DESCRIPTION
**Pull request by reta**
Signed-off-by: Andriy Redko <andriy.redko@aiven.io>

### Description
Implement SSL/TLS support for HTTP/2, see [1]

- HTTP/2 (default)
```
$ curl https://localhost:9200 -kv -u admin:admin

*   Trying 127.0.0.1:9200...
* Connected to localhost (127.0.0.1) port 9200 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* TLSv1.0 (OUT), TLS header, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS header, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS header, Finished (20):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Request CERT (13):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.2 (OUT), TLS header, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.3 (OUT), TLS handshake, Certificate (11):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_128_GCM_SHA256
* ALPN, server accepted to use h2
* Server certificate:
*  subject: DC=de; L=test; O=node; OU=node; CN=node-0.example.com
*  start date: Apr 22 03:43:47 2018 GMT
*  expire date: Apr 19 03:43:47 2028 GMT
*  issuer: DC=com; DC=example; O=Example Com Inc.; OU=Example Com Inc. Root CA; CN=Example Com Inc. Root CA
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
* Using HTTP2, server supports multiplexing
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* Server auth using Basic with user 'admin'
* Using Stream ID: 1 (easy handle 0x5653ef9e5e80)
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
> GET / HTTP/2
> Host: localhost:9200
> authorization: Basic YWRtaW46YWRtaW4=
> user-agent: curl/7.81.0
> accept: */*
> 
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* Connection state changed (MAX_CONCURRENT_STREAMS == 4294967295)!
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
< HTTP/2 200 
< content-type: application/json; charset=UTF-8
< content-length: 579
< 
{
  "name" : "runTask-0",
  "cluster_name" : "opensearch",
  "cluster_uuid" : "nf9IrHKAQb6VJtrLoR-4Xw",
  "version" : {
    "distribution" : "opensearch",
    "number" : "3.0.0-SNAPSHOT",
    "build_type" : "tar",
    "build_hash" : "ae6dbeaaad98c4849eff6c0a7863d9c751040b36",
    "build_date" : "2022-08-25T19:38:47.959594Z",
    "build_snapshot" : true,
    "lucene_version" : "9.4.0",
    "minimum_wire_compatibility_version" : "2.3.0",
    "minimum_index_compatibility_version" : "2.0.0"
  },
  "tagline" : "The OpenSearch Project: https://opensearch.org/"
}
```

- HTTP/1.1 

```
$ curl https://localhost:9200 -kv -u admin:admin --http1.1
*   Trying 127.0.0.1:9200...
* Connected to localhost (127.0.0.1) port 9200 (#0)
* ALPN, offering http/1.1
* TLSv1.0 (OUT), TLS header, Certificate Status (22):
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.2 (IN), TLS header, Certificate Status (22):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS header, Finished (20):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
* TLSv1.3 (IN), TLS handshake, Request CERT (13):
* TLSv1.3 (IN), TLS handshake, Certificate (11):
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
* TLSv1.3 (IN), TLS handshake, Finished (20):
* TLSv1.2 (OUT), TLS header, Finished (20):
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.3 (OUT), TLS handshake, Certificate (11):
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
* TLSv1.3 (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / TLS_AES_128_GCM_SHA256
* ALPN, server accepted to use http/1.1
* Server certificate:
*  subject: DC=de; L=test; O=node; OU=node; CN=node-0.example.com
*  start date: Apr 22 03:43:47 2018 GMT
*  expire date: Apr 19 03:43:47 2028 GMT
*  issuer: DC=com; DC=example; O=Example Com Inc.; OU=Example Com Inc. Root CA; CN=Example Com Inc. Root CA
*  SSL certificate verify result: unable to get local issuer certificate (20), continuing anyway.
* Server auth using Basic with user 'admin'
* TLSv1.2 (OUT), TLS header, Supplemental data (23):
> GET / HTTP/1.1
> Host: localhost:9200
> Authorization: Basic YWRtaW46YWRtaW4=
> User-Agent: curl/7.81.0
> Accept: */*
> 
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
* TLSv1.2 (IN), TLS header, Supplemental data (23):
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< content-type: application/json; charset=UTF-8
< content-length: 579
< 
{
  "name" : "runTask-0",
  "cluster_name" : "opensearch",
  "cluster_uuid" : "nf9IrHKAQb6VJtrLoR-4Xw",
  "version" : {
    "distribution" : "opensearch",
    "number" : "3.0.0-SNAPSHOT",
    "build_type" : "tar",
    "build_hash" : "ae6dbeaaad98c4849eff6c0a7863d9c751040b36",
    "build_date" : "2022-08-25T19:38:47.959594Z",
    "build_snapshot" : true,
    "lucene_version" : "9.4.0",
    "minimum_wire_compatibility_version" : "2.3.0",
    "minimum_index_compatibility_version" : "2.0.0"
  },
  "tagline" : "The OpenSearch Project: https://opensearch.org/"
}
```

[1] https://github.com/opensearch-project/OpenSearch/issues/3651

### Issues Resolved
- Closes https://github.com/opensearch-project/security/issues/2049
- Resolves #1953

### Testing
TODO: tests for HTTP/2

### Check List
- [ ] New functionality includes testing
- [ ] New functionality has been documented
- [X] Commits are signed per the DCO using --signoff

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by peternied — COMMENTED** [review]
Nice! Approach looks solid and straight forward.

You will see the BWC tests fail until https://github.com/opensearch-project/OpenSearch/issues/3615 has been resolved but that shouldn't stop you from making progress in the other GitHub Action issues

**Comment by reta** [comment]
> Nice! Approach looks solid and straight forward.
> 

Thanks a lot @peternied !

File: src/main/java/org/opensearch/security/ssl/DefaultSecurityKeyStore.java
Comment (cwperks) [inline src/main/java/org/opensearch/security/ssl/DefaultSecurityKeyStore.java:901]: I'm curious about this comment. I see that the same comment is present on the `Http2Client` example from netty: https://netty.io/4.1/xref/io/netty/example/http2/helloworld/client/Http2Client.html. Is the comment still true? In an effort to understand the default configuration here, would we ever consider using a different values for `SelectorFailureBehavior` and `SelectedListenerFailureBehavior`?

File: src/main/java/org/opensearch/security/ssl/DefaultSecurityKeyStore.java
Comment (cwperks) [inline src/main/java/org/opensearch/security/ssl/DefaultSecurityKeyStore.java:899]: Looks like this default configuration is repeated down below. Should we consider making this a static constant?

File: src/main/java/org/opensearch/security/ssl/DefaultSecurityKeyStore.java
Comment (reta) [reply src/main/java/org/opensearch/security/ssl/DefaultSecurityKeyStore.java:901]: > I'm curious about this comment. I see that the same comment is present on the Http2Client example from netty: https://netty.io/4.1/xref/io/netty/example/http2/helloworld/client/Http2Client.html. 

That is correct, it was picked up from there since we could use JDK and OpenSSL providers.

> Is the comment still true?

I suspect it is relevant since it is coming from Netty project.

> In an effort to understand the default configuration here, would we ever consider using a different values for SelectorFailureBehavior and SelectedListenerFailureBehavior?

Sure, seems to be a sensitive setting, we could always expose it for configuration if need arrives, thoughts?

File: src/main/java/org/opensearch/security/ssl/DefaultSecurityKeyStore.java
Comment (reta) [reply src/main/java/org/opensearch/security/ssl/DefaultSecurityKeyStore.java:899]: May be better as a function if we consider future possibility to configure `SelectorFailureBehavior` and `SelectedListenerFailureBehavior`?

File: src/main/java/org/opensearch/security/ssl/DefaultSecurityKeyStore.java
Comment (reta) [reply src/main/java/org/opensearch/security/ssl/DefaultSecurityKeyStore.java:901]: @cwperks quick update from the code: 
> Is the comment still true?
Yes, it is still true

Comment (cwperks) [reply src/main/java/org/opensearch/security/ssl/DefaultSecurityKeyStore.java:901]: @reta Thank you for checking.

File: src/main/java/org/opensearch/security/ssl/DefaultSecurityKeyStore.java
Comment (cwperks) [reply src/main/java/org/opensearch/security/ssl/DefaultSecurityKeyStore.java:899]: Sounds reasonable.

Comment (reta) [reply src/main/java/org/opensearch/security/ssl/DefaultSecurityKeyStore.java]: Done, thanks @cwperks !

**Review by peternied — DISMISSED** [review]
Thanks for the contribution, I am happy to merge these changes assuming the CI checks get fixed, please let us know if you need any help look into the CI failures.

**Comment by reta** [comment]
Thanks @peternied , looking into failures now

File: src/main/java/org/opensearch/security/ssl/DefaultSecurityKeyStore.java
Comment (DarshitChanpura) [inline src/main/java/org/opensearch/security/ssl/DefaultSecurityKeyStore.java:922]: Seems like the mentioned netty issue was fixed a while ago but this comment wasn't removed. Should we remove it? I know it maybe out of scope of this PR but wanted to know your thoughts

File: src/main/java/org/opensearch/security/ssl/util/SSLRequestHelper.java
Comment (DarshitChanpura) [inline src/main/java/org/opensearch/security/ssl/util/SSLRequestHelper.java:116]: For my curiosity, Can you help me understand why we are retrieving in sslHandler from an inboundPipeline() and why is this check required

Comment (reta) [reply src/main/java/org/opensearch/security/ssl/util/SSLRequestHelper.java:116]: @DarshitChanpura if you look into previous code, it looked at channel's pipeline to find out the SSL handler. Now, with HTTP/2, the new pipeline is created (hence, the inbound channel pipeline is "lost") and the SSL handler is not accessible anymore. The solution (sort of): keep inbound pipeline reference so the SSL handler could be retrieved.

**Comment by reta** [comment]
@peternied could you please help me with this Github action `CI / build-artifact-names` [1], I don't quite understand what it is doing by trying to locate next major OpenSearch version. Thank you.

[1] https://github.com/opensearch-project/security/actions/runs/3115185145/jobs/5051826328

**Comment by peternied** [comment]
@reta Nice work finding a bug in our version increment test cases, it was attempting to download maven artifacts from OpenSearch v4 which do not existing.  I've created #2106 to resolve this

**Comment by reta** [comment]
> Those BWC tests are stalling out this PR from merging, I've created a pull request #2108 to disable them, so this change _could_ update our the main branch to v3.0 as you've already done the hard work.

Thanks a lot @peternied , will try to help with BWC + 3.0 track

**Comment by peternied** [comment]
> will try to help with BWC + 3.0 track

Good news, BWC tests have started working with the 2.4.0 build, I've got https://github.com/opensearch-project/security/pull/2112 which has incorperated the fixes from this change into it.  This should rev the build number so your PR can focus around HTTP/2 support only.  Its been a long journey, nearly there...

**Comment by reta** [comment]
> > will try to help with BWC + 3.0 track
> 
> Good news, BWC tests have started working with the 2.4.0 build, I've got #2112 which has incorperated the fixes from this change into it. This should rev the build number so your PR can focus around HTTP/2 support only. Its been a long journey, nearly there...

Thanks a mill @peternied !

**Comment by reta** [comment]
:partying_face: thanks a lot for help @peternied !

**Comment by peternied** [comment]
@opensearch-project/security Could we get another review on this change?

**Review by DarshitChanpura — APPROVED** [review]
Ty @reta !

**Pull request merged**
